File: clojure-scripts/mason4.clj
(let [gridWidth 100
	  gridHeight 100

	  b_heptomino 	[[0 1 1]
		 			 [1 1 0]
		 			 [0 1 1]
		 			 [0 0 1]]

	  tutorial1 (proxy [sim.app.tutorial1and2.Tutorial1] [0]
		
		(start []
			(println "using overridden start()")
			;(proxy-super start)
			(set! (.grid this) (sim.field.grid.IntGrid2D. gridWidth gridHeight))
			(.seedGrid this)
			(.scheduleRepeating (.schedule this) (sim.app.tutorial1and2.CA.)))
			
		(seedGrid [] 
			(println "using overridden seedGrid()")
			(dotimes [x (count b_heptomino)] 
				(dotimes [y (count (nth b_heptomino x))] 
					(aset (.. this grid field) 
						  (- (+ x (/ (count (.. this grid field)) 2)) (/ (count b_heptomino) 2))
						  (- (+ y (/ (count (aget (.. this grid field) x)) 2)) (/ (count (nth b_heptomino x)) 2))
						  (int (nth (nth b_heptomino x) y)))))))]

	(defn display [field]
		(dotimes [y gridHeight]
			(dotimes [x gridWidth]
				(print (aget field x y)))
			(println)))

	(.start tutorial1)

	(while (< (.. tutorial1 schedule getSteps) 5000)
		(.. tutorial1 schedule (step tutorial1))
		(let [steps (.. tutorial1 schedule getSteps)]
			(if (= 0 (mod steps 500))
	        	(println (str "Steps: " steps " Time: " (.. tutorial1 schedule getTime))))))

	(display (.. tutorial1 grid field))

	(.finish tutorial1))

File: rhino-scripts/mason4.js

var MAX = 1000;

var mtime = function(fun) {
	var _start = java.lang.System.currentTimeMillis();
	fun();
	var _end = java.lang.System.currentTimeMillis();
	return (_end - _start) / 1000;
}

function setCell(agrid, x, y, value) {
	Packages.sim.scripting.ScriptHelper.set(agrid, x, y, value);
}
function lcount(agrid, x, y) {
	return Packages.sim.scripting.ScriptHelper.lcount(agrid, x, y);
}

var gridWidth = 100;
var gridHeight = 100;
var grid = new Packages.sim.field.grid.IntGrid2D(gridWidth, gridHeight); 
var tempGrid = new Packages.sim.field.grid.IntGrid2D(0, 0);

var b_heptomino = 	[[0, 1, 1], 
		 			 [1, 1, 0],
		 			 [0, 1, 1],
		 			 [0, 0, 1]];
		
var ca = new Packages.sim.engine.Steppable({
	step: function(state) {
		tempGrid.setTo(grid);
		for (var x = 0; x < gridWidth; x++) {
			for (var y = 0; y < gridHeight; y++) {
				var count = lcount(tempGrid, x, y);
				if (count <= 2 || count >= 5) {
					setCell(grid, x, y, 0);
				} else if (count == 3) {
					setCell(grid, x, y, 1);
				}
			}
		}
	}
});

var tutorial1 = new JavaAdapter(Packages.sim.engine.SimState, {
	start: function() {
		for (var x = 0; x < b_heptomino.length; x++) {
			for (var y = 0; y < b_heptomino[x].length; y++) {
				grid.field[Math.floor(x + grid.field.length / 2 + b_heptomino.length / 2)]
						  [Math.floor(y + grid.field[x].length / 2 + b_heptomino[x].length / 2)]
						=  b_heptomino[x][y];
			}
		}
		this.schedule.scheduleRepeating(ca);
	}
}, 0);
	
print(MAX / mtime(function() {
	tutorial1.start();

	while (tutorial1.schedule.getSteps() < MAX) {
		tutorial1.schedule.step(tutorial1);
		var steps = tutorial1.schedule.getSteps();
//		if (steps % 100 == 0) { print("Steps: " + steps); }
	}

	tutorial1.finish();
}));
